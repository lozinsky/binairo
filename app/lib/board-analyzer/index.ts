export * from './analyze-board-by-kind';
export * from './analyze-board';
