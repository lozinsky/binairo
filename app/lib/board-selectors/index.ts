export * from './select-filled-imbalanced-line';
export * from './select-line-that-can-be-equal-to-another';
export * from './select-line-that-includes-another';
export * from './select-line-that-is-equal-to-others';
export * from './select-middle-cell-between-two-identical';
export * from './select-next-cell-after-two-identical';
export * from './select-three-or-more-identical-sequential-cells';
export * from './select-unfilled-balanced-line';
