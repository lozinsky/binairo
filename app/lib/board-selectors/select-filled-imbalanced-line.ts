import type { Board } from '~/lib/board';

export function selectFilledImbalancedLine(target: Board) {
  for (const board of target.toDoubleRotation()) {
    for (const line of board) {
      if (line.isFilled && line.balance !== 0) {
        return { line, orientation: board.orientation };
      }
    }
  }

  return;
}
