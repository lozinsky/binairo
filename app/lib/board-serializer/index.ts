export * from './board-serializer-collection';
