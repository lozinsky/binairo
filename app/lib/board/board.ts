import {
  DoubleMatrixRotation,
  type Matrix,
  type MatrixReversible,
  type MatrixRotatable,
  type MatrixSelection,
  QuadrupleMatrixRotation,
} from '~/lib/matrix';
import { assert } from '~/shared/assert';
import { expectToBeDefined } from '~/shared/expect';

import type { BoardCell, BoardCellState } from './board-cell';

import { BoardLine, type BoardLineValue, isBoardLineValue } from './board-line';

export type BoardOrientation = 'landscape' | 'portrait';

const NEXT_BOARD_ORIENTATION_BY_PREV_BOARD_ORIENTATION: Readonly<Record<BoardOrientation, BoardOrientation>> = {
  landscape: 'portrait',
  portrait: 'landscape',
};

export type BoardValue = readonly BoardLineValue[];

export class Board implements Iterable<BoardLine>, Matrix<BoardLine>, MatrixReversible<Board>, MatrixRotatable<Board> {
  get length() {
    return this.#lines.length;
  }

  get orientation() {
    return this.#orientation;
  }

  get progress() {
    return this.#lines.reduce((progress, line) => progress + line.progress, 0) / this.#lines.length;
  }

  readonly #lines: readonly BoardLine[];
  readonly #orientation: BoardOrientation;

  constructor(orientation: BoardOrientation, lines: readonly BoardLine[]) {
    for (const line of lines) {
      assert(line.length === lines.length, 'Expected line length to be equal to board size');
    }

    this.#lines = lines;
    this.#orientation = orientation;
  }

  static blank(size: number) {
    return new this(
      'portrait',
      Array.from({ length: size }, () => BoardLine.blank(size)),
    );
  }

  static create(states: ReadonlyArray<readonly BoardCellState[]>) {
    return new this(
      'portrait',
      states.map((states) => BoardLine.create(states)),
    );
  }

  static from(value: BoardValue) {
    return new this(
      'portrait',
      value.map((value) => BoardLine.from(value)),
    );
  }

  at(index: number) {
    return this.#lines.at(index);
  }

  entries() {
    return this.#lines.entries();
  }

  replace(index: number, callback: (line: BoardLine) => BoardLine) {
    return new Board(this.#orientation, this.#lines.with(index, callback(expectToBeDefined(this.#lines.at(index)))));
  }

  replaceBy(selection: MatrixSelection, callback: (cell: BoardCell) => BoardCell) {
    const content = selection.execute(this);
    const lines = this.#lines.map((line) => {
      const cells = Array.from(line);

      for (const cell of content) {
        const index = line.indexOf(cell);

        if (index === -1) {
          continue;
        }

        cells[index] = callback(expectToBeDefined(cells[index]));
      }

      return new BoardLine(cells);
    });

    return new Board(this.#orientation, lines);
  }

  reverse() {
    return new Board(this.#orientation, this.#lines.toReversed());
  }

  rotate() {
    const orientation = NEXT_BOARD_ORIENTATION_BY_PREV_BOARD_ORIENTATION[this.#orientation];

    if (this.#lines.length === 0) {
      return new Board(orientation, []);
    }

    const lines: BoardLine[] = [];

    for (const index of expectToBeDefined(this.#lines[0]).keys()) {
      const cells: BoardCell[] = [];

      for (const line of this.#lines) {
        cells.push(expectToBeDefined(line.at(index)));
      }

      lines.push(new BoardLine(cells));
    }

    return new Board(orientation, lines);
  }

  slice(start: number, end?: number) {
    return this.#lines.slice(start, end);
  }

  [Symbol.iterator]() {
    return this.#lines[Symbol.iterator]();
  }

  toDoubleRotation() {
    return DoubleMatrixRotation.run<Board>(this);
  }

  toQuadrupleRotation() {
    return QuadrupleMatrixRotation.run<Board>(this);
  }

  valueOf(): BoardValue {
    return this.#lines.map((line) => line.valueOf());
  }
}

export function isBoardValue(value: unknown): value is BoardValue {
  return Array.isArray(value) && value.every(isBoardLineValue);
}
