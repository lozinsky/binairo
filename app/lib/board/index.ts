export * from './board-cell';
export * from './board-line';
export * from './board';
