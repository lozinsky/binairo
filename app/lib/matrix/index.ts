export * from './matrix-line';
export * from './matrix-rotation';
export * from './matrix-selection';
export * from './matrix';
