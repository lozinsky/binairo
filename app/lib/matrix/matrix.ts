import type { MatrixLine, MatrixLineCell } from './matrix-line';

export interface Matrix<L extends MatrixLine<unknown>> {
  at(index: number): L | undefined;
  entries(): Iterable<[number, L]>;
}

export type MatrixCell<M> = M extends Matrix<infer L> ? MatrixLineCell<L> : never;
