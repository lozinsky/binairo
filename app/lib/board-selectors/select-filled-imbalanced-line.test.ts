import { assert, expect, test } from 'vitest';

import { Board } from '~/lib/board';
import { MatrixSelection } from '~/lib/matrix';

import { selectFilledImbalancedLine } from './select-filled-imbalanced-line';

test.each([
  Board.create([
    ['E', 'E', 'E', 'E'],
    ['R', 'R', 'R', 'B'],
    ['E', 'E', 'E', 'E'],
    ['E', 'E', 'E', 'E'],
  ]),
  Board.create([
    ['E', 'E', 'E', 'E'],
    ['R', 'R', 'R', 'R'],
    ['E', 'E', 'E', 'E'],
    ['E', 'E', 'E', 'E'],
  ]),
  Board.create([
    ['E', 'R', 'E', 'E'],
    ['E', 'R', 'E', 'E'],
    ['E', 'R', 'E', 'E'],
    ['E', 'B', 'E', 'E'],
  ]),
  Board.create([
    ['E', 'R', 'E', 'E'],
    ['E', 'R', 'E', 'E'],
    ['E', 'R', 'E', 'E'],
    ['E', 'R', 'E', 'E'],
  ]),
])('selects filled imbalanced line', (target) => {
  const payload = selectFilledImbalancedLine(target);

  assert(payload !== undefined);

  const selection = MatrixSelection.collect(target, Array.from(payload.line));

  expect(payload.orientation).toMatchSnapshot();
  expect(target.replaceBy(selection, (cell) => cell.toFixed())).toMatchSnapshot();
});

test.each([
  Board.create([
    ['E', 'E', 'E', 'E'],
    ['R', 'R', 'B', 'B'],
    ['E', 'E', 'E', 'E'],
    ['E', 'E', 'E', 'E'],
  ]),
  Board.create([
    ['E', 'E', 'E', 'E'],
    ['R', 'R', 'E', 'E'],
    ['E', 'E', 'E', 'E'],
    ['E', 'E', 'E', 'E'],
  ]),
  Board.create([
    ['E', 'R', 'E', 'E'],
    ['E', 'R', 'E', 'E'],
    ['E', 'B', 'E', 'E'],
    ['E', 'B', 'E', 'E'],
  ]),
  Board.create([
    ['E', 'R', 'E', 'E'],
    ['E', 'R', 'E', 'E'],
    ['E', 'E', 'E', 'E'],
    ['E', 'E', 'E', 'E'],
  ]),
])('returns "undefined" if no filled imbalanced line exist', (target) => {
  expect(selectFilledImbalancedLine(target)).toMatchSnapshot();
});
