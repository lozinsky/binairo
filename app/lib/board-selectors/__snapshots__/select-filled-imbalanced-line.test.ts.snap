// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`returns "undefined" if no filled imbalanced line exist 1`] = `undefined`;

exports[`returns "undefined" if no filled imbalanced line exist 2`] = `undefined`;

exports[`returns "undefined" if no filled imbalanced line exist 3`] = `undefined`;

exports[`returns "undefined" if no filled imbalanced line exist 4`] = `undefined`;

exports[`selects filled imbalanced line 1`] = `"portrait"`;

exports[`selects filled imbalanced line 2`] = `
[⬜] [⬜] [⬜] [⬜]
{🟥} {🟥} {🟥} {🟦}
[⬜] [⬜] [⬜] [⬜]
[⬜] [⬜] [⬜] [⬜]
`;

exports[`selects filled imbalanced line 3`] = `"portrait"`;

exports[`selects filled imbalanced line 4`] = `
[⬜] [⬜] [⬜] [⬜]
{🟥} {🟥} {🟥} {🟥}
[⬜] [⬜] [⬜] [⬜]
[⬜] [⬜] [⬜] [⬜]
`;

exports[`selects filled imbalanced line 5`] = `"landscape"`;

exports[`selects filled imbalanced line 6`] = `
[⬜] {🟥} [⬜] [⬜]
[⬜] {🟥} [⬜] [⬜]
[⬜] {🟥} [⬜] [⬜]
[⬜] {🟦} [⬜] [⬜]
`;

exports[`selects filled imbalanced line 7`] = `"landscape"`;

exports[`selects filled imbalanced line 8`] = `
[⬜] {🟥} [⬜] [⬜]
[⬜] {🟥} [⬜] [⬜]
[⬜] {🟥} [⬜] [⬜]
[⬜] {🟥} [⬜] [⬜]
`;
